### PR TITLE
healthcheck: add regular chain backend and disk space checks

### LINKED ISSUE #1557: [Feature Request] Cookie file authentication should not be cached
I noticed you are caching cookiefile at startup by considering it the same as RPCUser/Pass in `extractBitcoindParameters`. But the cookiefile change when BitcoinD get restarted, which mean LND stop working if bitcoind restart.

Second issue with it, if LND start just after BitcoinD but just before the cookiefile get regenerated, then we start having problem because LND is using the outdated file.

This make integration of LND to BTCPay randomly working.

### LINKED ISSUE #3109: [Bug] Make LND resilient to bitcoind crashes: Dropping ZMQ connection
### Background

Very often, when BitcoinD restarts for anything, LND stop working.

I tracked that previously to LND not handling cookiefile authentication the right way.
LND not reading the cookiefile everytimes it is calling RPC. **BUT this is not the only problem.**

On BTCPay, I switched to fixed rpcuser/password only for LND, but keep having problems.

The last time I noticed that my LND was seemingly working, but the chain was not synched anymore for some reason.
I checked the logs, could not find any specific issue, but it seemed clear LND was not getting blocks from BitcoinD, and it was also clear that LND was able to use RPC against bitcoind  (no RPC error in the logs).

Restarting LND did the trick.

### Your environment

* version of `lnd 0.6.1`

### Steps to reproduce

* Run Bitcoin Core
* Run LND
* Wait LND is operational
* Restart Bitcoin Core

### Expected behaviour

LND should reconnect to ZMQ automatically OR crash.

### Actual behaviour

LND does not reconnect to ZMQ, the chain is not synched anymore, the blocks are not received, putting the user's fund at risk.

### PR DESCRIPTION
**Pull request by carlaKC**
This PR wraps our existing logger in one which will request graceful shutdown when the `Critical` logging level is used. This wrapper is used to add two health checks to lnd which will shutdown if we do not succeed after sufficient backoff/retry:
1. Check that we have a connection to our chain backend
2. Check that we have some space available on disk

The regularity and strictness (ie, number of retries we allow) of this check are made configurable, so that users can turn them off/ make them less strict, or turn them off entirely if they cause issues (particularly thinking about neutrino flakes).

Windows is windows, so we need to have a special windows space check (`_windows`, TIL!). 

Partially addresses #4306 (does not cover failed disk writes, so issue not entirely fixed)

Fixes #1557
Fixes #3109

### CODE REVIEW

File: build/log_shutdown.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		shutdown: shutdown,
+	}
+}
+
+// Criticalf formats message according to format specifier and writes to
+// log with LevelCritical. It will then call the shutdown logger's shutdown
+// function to prompt safe shutdown.
+//
+// Note: is it part of the btclog.Logger interface.
+func (s *ShutdownLogger) Criticalf(format string, params ...interface{}) {
```
Comment (Roasbeef) [inline build/log_shutdown.go:46]: 👍

File: build/log_shutdown.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	// shutdown is the shutdown function we will use to signal shutdown.
+	shutdown func()
+}
+
+// NewShutdownLogger creates a shutdown logger for the log provided which will
+// use the signal package to request shutdown on critical errors.
+func NewShutdownLogger(logger btclog.Logger) *ShutdownLogger {
+	shutdown := func() {
+		// If our application is not listening for interrupts, we return
```
Comment (Roasbeef) [inline build/log_shutdown.go:21]: Can't this just be a regular method body given it closes over the logger which is already embedded as an attribute in the struct?

File: signal/signal.go [hunk 3]
```diff
@@ -45,6 +48,10 @@ func Intercept() {
 // callback registration.
 // It must be run as a goroutine.
 func mainInterruptHandler() {
+	// Close our started channel to signal that we have started listening
+	// for interrupts.
+	close(started)
```
Comment (Roasbeef) [inline signal/signal.go:53]: This would cause a double close panic if `Intercept` is ever called twice....

Perhaps we should try to enforce that there is only a singleton registration for interrupt signals?

File: config.go [hunk 4]
```diff
@@ -86,6 +86,10 @@ const (
 	// HostAnnouncer will wait between DNS resolutions to check if the
 	// backing IP of a host has changed.
 	defaultHostSampleInterval = time.Minute * 5
+
+	defaultChainInterval = time.Minute * 10
```
Comment (Roasbeef) [inline config.go:90]: Hmm, this seems a bit too high...I'd imagine we'd want to check at least once a minute, just so any hypervisor-like monitor can restart `lnd` shortly after the backend shuts down.

File: healthcheck/healthcheck.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
+// Package healthcheck contains takes contains a monitor which takes a set
```
Comment (Roasbeef) [inline healthcheck/healthcheck.go:1]: 👍

File: config.go (hunk 4 above)
Comment (Roasbeef) [reply config.go:90]: An alternative to polling would be to rely on the `zmq` connection for `bitcoind` and the websockets connection for `btcd`. If either of these connections fail (EOF error), then we know the backend is down.

File: healthcheck/healthcheck.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	}
+}
+
+// retryCheck calls a check function until it succeeds, or we exceed our
+// configured number of retries. If we fail to obtain a passing health check
+// after the configured number of retries, we will request safe shutdown.
+func retryCheck(check *Observation, quit chan struct{}, shutdown shutdownFunc) {
+	var count int
+	for count < check.Retries {
+		err := check.Check()
```
Comment (Roasbeef) [inline healthcheck/healthcheck.go:144]: if retries is zero (lnd should shutdown if the check fails even once), then this loop is never executed. We want more of a "do -- while" loop here.

File: healthcheck/healthcheck_test.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		expectedCalls int
+
+		// expectedShutdown is true if we expect a shutdown to be
+		// triggered because all of our calls failed.
+		expectedShutdown bool
+	}{
+		{
+			name:             "first call succeeds",
+			callErrors:       []error{nil},
+			retries:          2,
```
Comment (Roasbeef) [inline healthcheck/healthcheck_test.go:145]: Should add a test case for zero retries.

File: server.go [hunk 8]
```diff
@@ -1240,6 +1244,34 @@ func newServer(cfg *Config, listenAddrs []net.Addr, chanDB *channeldb.DB,
 		})
 	}
 
+	var healthChecks []*healthcheck.Observation
+	if !cfg.HealthChecks.ChainCheck.Disabled {
+		backendCheck := func() error {
+			_, _, err := cc.chainIO.GetBestBlock()
```
Comment (Roasbeef) [inline server.go:1250]: Need to double check how this degrades for each backend...I think for `btcd` this might just block indefinitely rather than actually returning immediately with an error.

File: config.go (hunk 4 above)
Comment (carlaKC) [reply config.go:90]: Relying on these connections meaning make a change in `btcwallet` so that we can access the existing connection, or add a direct connection to zmq/websockets from the health check?

**Comment by carlaKC** [comment]
Rebased and updated in separate pushes, diff of changes [here](https://github.com/lightningnetwork/lnd/compare/42ee2696779367db67ad554b87badd86d9281295..e776d99ec1277fbe8a6df0797119583b7d00b9fd). Changes since last review:
- Update changes to signal package so that we can only call intercept once
- Add timeout to health checks (because btcd `GetBestBlock` blocks when btcd shuts down)
- Add disk check for solaris behind build tag, as done for windows

File: build/log_shutdown.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+func (s *ShutdownLogger) Criticalf(format string, params ...interface{}) {
+	s.Logger.Criticalf(format, params...)
+	s.shutdown()
+}
+
+// Critical formats message using the default formats for its operands
+// and writes to log with LevelCritical. It will then call the shutdown
+// logger's shutdown function to prompt safe shutdown.
+//
+// Note: is it part of the btclog.Logger interface.
```
Comment (Crypt-iQ) [inline build/log_shutdown.go:36]: nit: It is

File: healthcheck/diskcheck_solaris.go [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+func AvailableDiskSpace(path string) (float64, error) {
+	s := unix.Stat_t{}
+	err := unix.Stat(path, &s)
+	if err != nil {
+		return 0, err
+	}
+
+	// Calculate our free blocks/total blocks to get our
+	// total number of free blocks.
+	return float64(s.Blocks) / float64(s.Blocks), nil
```
Comment (Crypt-iQ) [inline healthcheck/diskcheck_solaris.go:16]: `s.Blocks` -> `s.Bfree` ?

File: config.go (hunk 4 above)
Comment (Roasbeef) [reply config.go]: Thinking just a direction connection so we don't need to dig down into the wallet. For `bitcoind`, we can only request the block notifications which aren't as spammy. A similar option can be applied to the `btcd` web sockets connection.

**Comment by Roasbeef** [comment]
> Is it possible to make integration tests out of these with the various backends and kill the backends to see that lnd does shutdown? I don't think it would be possible to really test the disk healthcheck. 

That's possible, but the test set up would need to be slightly distinct as all the other tests assumes that the target chain backend will be around the entire time. Agreed about the disk check....one way to test it would be using a container w/ a very small volume (would basically be unable to start).

File: lncfg/healthcheck.go [hunk 11]
```diff
@@ -23,11 +24,27 @@ var (
...
+	}
+
+	if err := h.DiskCheck.validate("disk space"); err != nil {
+		return err
+	}
+
+	if h.DiskCheck.RequiredRemaining < 0 ||
+		h.DiskCheck.RequiredRemaining >= 1 {
+
+		return errors.New("disk required ratio must be in [0;1)")
```
Comment (Crypt-iQ) [inline lncfg/healthcheck.go:44]: `[0:1)` maybe?

**Comment by carlaKC** [comment]
> Haven't been able to test this out

I have run some local tests for disk check, and chain backend for btcd + bitcoind, so it depends on how much time we want to put into testing. I can wrangle together some bash scripts to put up and kill docker containers if needed. 

> `btcd` hang comment above

Right now `btcd` does hang, which is why the timeout was added for these checks. The unfortunate part is that the goroutine calling the best block endpoint will hang indefinitely, but there's not much that can be done unless we make a change to btcwallet. There is the option of using zmq/websockets but that comes with some copy-fu from btcwallet, and halves the number of lnd nodes bitcoind can serve (down to 2, which sucks for local dev).

**Comment by Crypt-iQ** [comment]
> > Haven't been able to test this out
> 
> I have run some local tests for disk check, and chain backend for btcd + bitcoind, so it depends on how much time we want to put into testing. I can wrangle together some bash scripts to put up and kill docker containers if needed.
> 
> > `btcd` hang comment above
> 
> Right now `btcd` does hang, which is why the timeout was added for these checks. The unfortunate part is that the goroutine calling the best block endpoint will hang indefinitely, but there's not much that can be done unless we make a change to btcwallet. There is the option of using zmq/websockets but that comes with some copy-fu from btcwallet, and halves the number of lnd nodes bitcoind can serve (down to 2, which sucks for local dev).

Would be super helpful to have some docker scripts, maybe could even be run as part of our tests?
